tuf/tuf.go

```diff
@@ -173,6 +173,112 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
+	if !data.IsBaseRole(name) {
+		return data.BaseRole{}, data.ErrInvalidRole{Role: name, Reason: "invalid base role name"}
+	}
+	if tr.Root == nil {
+		return data.BaseRole{}, ErrNotLoaded{data.CanonicalRootRole}
+	}
+	roleData, ok := tr.Root.Signed.Roles[name]
+	if !ok {
+		return data.BaseRole{}, ErrNotLoaded{name}
+	}
+	// Get all public keys for the base role from TUF metadata
+	keyIDs := roleData.KeyIDs
+	pubKeys := make(map[string]data.PublicKey)
+	for _, keyID := range keyIDs {
+		pubKey, ok := tr.Root.Signed.Keys[keyID]
+		if !ok {
+			return data.BaseRole{}, fmt.Errorf("key with ID %s, specified as a signing key for role %s, was not found in root metadata", keyID, name)
+		}
+		pubKeys[keyID] = pubKey
+	}
+
+	return data.BaseRole{
+		Name:      name,
+		Keys:      pubKeys,
+		Threshold: roleData.Threshold,
+	}, nil
+}
+
+// GetDelegationRole gets a delegation role from this repo's metadata, walking from the targets role down to the delegation itself
+func (tr *Repo) GetDelegationRole(name string) (data.DelegationRole, error) {
+	if !data.IsDelegation(name) {
+		return data.DelegationRole{}, data.ErrInvalidRole{Role: name, Reason: "invalid delegation name"}
+	}
+	if tr.Root == nil {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalRootRole}
+	}
+	_, ok := tr.Root.Signed.Roles[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalTargetsRole}
+	}
+	// Traverse target metadata, down to delegation itself
+	// Get all public keys for the base role from TUF metadata
+	signedTargetData, ok := tr.Targets[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, fmt.Errorf("TUF data for targets does not exist in targets metadata", name)
+	}
+
+	// Keep track of paths and path hash prefixes to restrict as we traverse the delegations tree, targets implicitly has ""
+	whiteListedPaths := []string{""}
+	whiteListedPathHashes := []string{""}
+
+	// Start with top level roles in targets
+	delegationRoles := signedTargetData.Signed.Delegations.Roles
+	var foundRole *data.Role
+	for len(delegationRoles) > 0 {
+		delgRole := delegationRoles[0]
+		delegationRoles = delegationRoles[1:]
+
+		// If this role is delegated above or is our desired role, restrict paths and traverse its child roles
+		if delgRole.Name == name || strings.HasPrefix(name, delgRole.Name+"/") {
+			delgRole.Paths = data.RestrictDelegationPathPrefixes(whiteListedPaths, delgRole.Paths)
+			delgRole.PathHashPrefixes = data.RestrictDelegationPathPrefixes(whiteListedPathHashes, delgRole.PathHashPrefixes)
+
+			// If we found the role, we can exit the loop
+			if delgRole.Name == name {
+				foundRole = delgRole
+				break
+			}
+
+			// If this is a parent role, keep traversing
+			whiteListedPaths = delgRole.Paths
+			whiteListedPathHashes = delgRole.PathHashPrefixes
+			delegationRoles = append(delegationRoles, tr.Targets[delgRole.Name].Signed.Delegations.Roles...)
```

Comment:
We check that the parent role's metadata must exist in later, on [https://github.com/docker/notary/pull/566/files#diff-b7b7bae92e6be55133dab8ee3256e014R260](line 260), but shouldn't we also check here?  If the metadata doesn't exist for the role, will this panic?